### ISSUE TITLE
Parts: Wrong location for parts

### ISSUE THREAD

**Issue by eibre** [Autodesk/revit-ifc#86]
Version: IFC4RV Beta Release v0.6 .2

## Steps to reproduce:
I have a simple I-beam where fire insultion is added.
Create parts of the beam
Export ifc in a view where parts are visible.
## Problem:
The position of the parts are offseted.
If "export parts as building elements" is checked, the position seems to be correct.
![image](https://user-images.githubusercontent.com/5968613/51600745-f8700f80-1f02-11e9-81ab-d469b7583e83.png)

Sample Revit and two ifc files created with different settings are attached:
[sample_files.zip](https://github.com/Autodesk/revit-ifc/files/2786787/sample_files.zip)
Maybe this is related to https://github.com/Autodesk/revit-ifc/issues/20

**Comment by WawanSolihin** [comment]
In this case, what is your expectation on the exported IFC? Is the beam supposed to be created as a Beam with aggregation of the parts? or it is OK with a Beam with multiple geometries?
It seems that in the latest version, the later (Beam with multiple geometry items are created). The position seems to be correct.

**Comment by dvrvb** [comment]
With a 'standard/default' export (without the option of 'Export parts as building elements' ticked), it is now creating the aggregation: I think that's OK.
<img width="331" alt="image" src="https://user-images.githubusercontent.com/41319048/51849987-1a153080-2321-11e9-87f5-3eedd12e85d8.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/41319048/51850248-ac1d3900-2321-11e9-9afd-7cc86854c3bd.png">

But in that case the parts are running away.

**Comment by WawanSolihin** [comment]
Maybe I was not very clear. I tested the file with the latest ifcexporter and I do not get a IfcBeam with an aggregate of 5 IfcBuildingElementParts, but I get an IfcBeam that has 5 geometry items instead. The location seems to be correct (see attached). I don't think this change in the behaviour is on purpose, but it is a side effect of supporting the creation of the Type object (e.g. IfcBeamType in this case) that will have a shared geometry (using MappedItem), which is shared by multiple instances of the same type. In the past the type was not supported, or not real (i.e. one type for each instance).

So my question here is which is the desired behavior that is preferred? An IfcBeam with aggregate of 5 IfcBuildingElementPart, or an IfcBeam with a reference to IfcBeamType that has 5 geometry items (linked to the instance by MappedItem)?

**Comment by WawanSolihin** [comment]
The attachment:
[beam-part-sampleRV.zip](https://github.com/Autodesk/revit-ifc/files/2804103/beam-part-sampleRV.zip)

**Comment by dvrvb** [comment]
with the DEV-solution until 21th january (not your latest commit of 25 jan) I got this result, with the aggregation: see sample
[beam-part-sample_dvrvb.zip](https://github.com/Autodesk/revit-ifc/files/2804133/beam-part-sample_dvrvb.zip)

**Comment by WawanSolihin** [comment]
Yup I got that at first until I updated to the latest codes. I do not think there was any particular change directly related to it between 21 Jan to now. 
I need to take a closer look.

**Comment by dvrvb** [comment]
I'm building now the latest solution. I thought it was only related to some German translations, but I will notice it in a few moments.

**Comment by dvrvb** [comment]
Even with the latest commit, I still got the aggregation.
Can you check if you export it with:
<img width="303" alt="image" src="https://user-images.githubusercontent.com/41319048/51855399-767e4d00-232d-11e9-970d-a0d5266c8960.png">

**Comment by WawanSolihin** [comment]
Hmm that's weird. I used the file provided above, no other change made:
![image](https://user-images.githubusercontent.com/31854582/51855607-9c751300-2368-11e9-9ae0-0cf143fa75b6.png)

**Comment by dvrvb** [comment]
Aha, if I choose the default export option IFC4 Reference View, I also got the same result as you.
But if I use 'my default IFC4 RV json', I will get the aggregation.
[IFC_Configuration_-_IFC4RV1.2.json.zip](https://github.com/Autodesk/revit-ifc/files/2804231/IFC_Configuration_-_IFC4RV1.2.json.zip)

**Comment by WawanSolihin** [comment]
Hmm true enough, that is the case. There is a setting there that makes the import parts somewhere. I will take a look.

**Comment by dvrvb** [comment]
<img width="213" alt="image" src="https://user-images.githubusercontent.com/41319048/51856461-0de49f80-2330-11e9-84ad-4938ffde760c.png">
It seems related to this setting.

**Comment by eibre** [comment]
Yes @dvrvb is right. Parts are not exported unless `Export only elements visible in view` is checked and `show parts` is chosen in that views propery.

**Comment by WawanSolihin** [comment]
1. Parts will only be exported as an aggregation when the 2 settings are set. This is an "as-designed" behavior as I understand it.
2. Commit 958482e fixes the issue of incorrect offset for parts

**Comment by eibre** [comment]
@WawanSolihin I installed the last[ beta5](https://github.com/Autodesk/revit-ifc/releases/tag/Beta5-v19.2.0) and exported with the settings shown in the image below.
![image](https://user-images.githubusercontent.com/5968613/53148011-98c05f00-35a9-11e9-8bad-a414c408a3e6.png)

This is how the result looks like;
![image](https://user-images.githubusercontent.com/5968613/53148025-a70e7b00-35a9-11e9-83e4-00b4aed020b4.png)

**Comment by dvrvb** [comment]
Hi Eibre,

I don't want to interfere in your question at Wawan.
But although I also tested your sample before I closed the similar issue at https://github.com/Autodesk/revit-ifc/issues/20
I'm now curious how your test differs from mine.
Could you test the sample in the attached zip ? (it is your beam, placed on a floor, to monitor if it stays in place or not).
[2019-02-21.zip](https://github.com/Autodesk/revit-ifc/files/2888073/2019-02-21.zip)

If I export it now (IFC2x3CV2.0 and IFC4RV1.2), with a current build solution, it still stays in place.
So I'm wondering what the result is when you export it.

Regards,
Dirk 
<img width="796" alt="image" src="https://user-images.githubusercontent.com/41319048/53150174-39198200-35b0-11e9-9e7d-1fa7cccc993c.png">

**Comment by eibre** [comment]
@dvrvb Thank you for your interest in this issue! I tested your file and the parts stay in place, the same result as you got. How ever when I added a true north rotation to project base point they are off again:
![image](https://user-images.githubusercontent.com/5968613/53264142-3a939900-36db-11e9-984e-dea246a0d1a3.png)
![image](https://user-images.githubusercontent.com/5968613/53264168-497a4b80-36db-11e9-8241-6f28bded5172.png)

**Comment by WawanSolihin** [comment]
Thanks. That's important clue. I will check.

Regards,
Wawan

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by WawanSolihin** [comment]
Hi @eibre, 
The issue is more complicated than it appears. I need to make change that might cause regression to fic this in the right way. Do you have more test cases that contain elements with and without Parts to perform more test? Or if not possible, I can also make test version for you to check on models that you may not be able to share.

Regards,
Wawan

**Comment by eibre** [comment]
@WawanSolihin I can't share my model to the public, but have sendt it to Angel since I couldn't find your e-mail address.

**Comment by ArtemHnatko** [comment]
The issue described with parts having an incorrect position during IFC export does not reproduce in Revit 2023. Both scenarios, with and without "Export parts as building elements" and "Export only elements visible in view" enabled, produce correct part positions in the exported IFC.
**Testing environment:**
- Autodesk Revit 2023 (package: 2023_Ship_x64_PR_20240924_0423_9a3baabb0895)
- Autodesk Revit 2020 (package: 2020_Ship_x64_CI_20230110_0827_056a579da61a) 

![image](https://github.com/user-attachments/assets/4639dfe4-a601-4057-9cbe-7680dac2e684)
